File: ios/sdk/swift/internal/ContactListener.swift
import CrossPL

/* @CrossClass */
@objc open class ContactListener: CrossBase {
  open func onError(errCode: Int32, errStr: String, ext: String?) {
    fatalError("\(#function) not implementation.")
  }
  open func onAcquire(request: AcquireArgs) -> Data? {
    fatalError("\(#function) not implementation.")
  }
  open func onEvent(event: EventArgs) {
    fatalError("\(#function) not implementation.")
  }
  open func onReceivedMessage(humanCode: String, channelType: Contact.Channel, message: Contact.Message) {
    fatalError("\(#function) not implementation.")
  }
  
  public typealias AcquireArgs = ContactInternal.AcquireArgs
  public typealias EventArgs = ContactInternal.EventArgs

  public class StatusEvent: EventArgs {
    public override init(type: Int, humanCode: String, channelType: Int, data: Data?) {
      status = Contact.Status(rawValue: Int(data![0]))!
      super.init(type: type, humanCode: humanCode, channelType: channelType, data: data)
    }
   
    public override func toString() -> String {
      return  "StatusEvent"
            + "[type=\(type)"
            + ",humanCode=\(humanCode)"
            + ",channelType=\(channelType)"
            + ",status=\(status)"
            + "]"
    }

    public let status: Contact.Status
  }

  public class RequestEvent: EventArgs {
    public override init(type: Int, humanCode: String, channelType: Int, data: Data?) {
      summary = Data.ToString(from: data) ?? ""
      super.init(type: type, humanCode: humanCode, channelType: channelType, data: data)
    }
    
    public override func toString() -> String {
    return  "StatusEvent"
          + "[type=\(type)"
          + ",humanCode=\(humanCode)"
          + ",channelType=\(channelType)"
          + ",summary=\(summary)"
          + "]"
    }

    public let summary: String
  }

  public class InfoEvent: EventArgs {
    public override init(type: Int, humanCode: String, channelType: Int, data: Data?) {
      let info =  Data.ToString(from: data) ?? ""
      if info.contains(ContactInternal.JsonKey.IsMyself) == true {
        let userInfo = Contact.UserInfo()
        _ = userInfo.fromJson(info: info)
        humanInfo = userInfo
      } else if info.contains(ContactInternal.JsonKey.IsFriend) == true {
        let friendInfo = Contact.FriendInfo()
        _ = friendInfo.fromJson(info: info)
        humanInfo = friendInfo
      } else {
        ContactInternal.Log.w(tag: Contact.TAG, msg: "InfoEvent: Failed to parse human data.");
      }
      
      humanInfo = Contact.HumanInfo()
      super.init(type: type, humanCode: humanCode, channelType: channelType, data: data)
    }

    public override func toString() -> String {
      return  "StatusEvent"
            + "[type=\(type)"
            + ",humanCode=\(humanCode)"
            + ",channelType=\(channelType)"
            + ",humanInfo=\(humanInfo.toString())"
            + "]"
    }
    
    public private(set) var humanInfo: Contact.HumanInfo
  }

  public init() {
    super.init(className: String(describing: ContactListener.self))
  }

  /* @CrossPlatformInterface */
  @objc internal func onAcquire(_ reqType: Int, _ pubKey: String?, _ data: Data?) -> Data? {
    ContactInternal.Log.i(tag: Contact.TAG, msg: "ContactListener.onAcquire()")
//
    let args = AcquireArgs(type: reqType, pubKey: pubKey, data: data)
    let ret = onAcquire(request: args);

    return ret;
  }

  /* @CrossPlatformInterface */
  @objc internal func onEvent(_ eventType: Int, _ humanCode: String, _ channelType: Int, _ data: Data) {

    var args: EventArgs

    let type = EventArgs.Kind(rawValue: eventType)
    switch (type) {
    case .StatusChanged:
      args = StatusEvent(type: eventType, humanCode: humanCode, channelType: channelType, data: data)
      break;
    case .FriendRequest:
      args = RequestEvent(type: eventType, humanCode: humanCode, channelType: channelType, data: data)
      break;
    case .HumanInfoChanged:
      args = InfoEvent(type: eventType, humanCode: humanCode, channelType: channelType, data: data)
      break;
    default:
      fatalError("Unimplemented type: \(String(describing: type))");
    }

    ContactInternal.Log.i(tag: Contact.TAG, msg: "ContactListener.onEvent() args=\(args.toString())");
    onEvent(event: args);
    return;
  }

  /* @CrossPlatformInterface */
  @objc internal func onReceivedMessage(_ humanCode: String, _ channelType: Int,
                                       _ type: Int, _ data: Data,
                                       _ cryptoAlgorithm: String, _ timestamp: Int64) {
    let message = Contact.Message(type: Contact.Message.Kind(rawValue: type)!,
                                  data: data,
                                  cryptoAlgorithm: cryptoAlgorithm)
    message.timestamp = timestamp

    onReceivedMessage(humanCode: humanCode,
                      channelType: Contact.Channel(rawValue: channelType)!,
                      message: message);
    return;
  }
  
  /* @CrossPlatformInterface */
  @objc internal func onError(_ errCode: Int32, _ errStr: String, _ ext: String?) {
    onError(errCode: errCode, errStr: errStr, ext: ext)
  }
}

File: android/sdk/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/ContactListener.java
package org.elastos.sdk.elephantwallet.contact.internal;

import android.util.Log;

import com.google.gson.JsonObject;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.elastos.sdk.elephantwallet.contact.Utils;
import org.elastos.tools.crosspl.CrossBase;
import org.elastos.tools.crosspl.annotation.CrossClass;
import org.elastos.tools.crosspl.annotation.CrossInterface;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

@CrossClass
abstract class ContactListener extends CrossBase {
    @CrossInterface
    public abstract void onError(int errCode, String errStr, String ext);

    public abstract byte[] onAcquire(AcquireArgs request);
    public abstract void onEvent(EventArgs event);
    public abstract void onReceivedMessage(String humanCode, Contact.Channel channelType, Contact.Message message);

    public class AcquireArgs extends org.elastos.sdk.elephantwallet.contact.internal.AcquireArgs {
        private AcquireArgs(int type, String pubKey, byte[] data, String extra) {
            super(type, pubKey, data, extra);
        }
    }

    public class EventArgs extends org.elastos.sdk.elephantwallet.contact.internal.EventArgs {
        public EventArgs(int type, String humanCode, int channelType, byte[] data) {
            super(type, humanCode, channelType, data);
        }
    }

    public class StatusEvent extends EventArgs {
        public StatusEvent(int type, String humanCode, int channelType, byte[] data) {
            super(type, humanCode, channelType, data);
            status = Contact.Status.valueOf(data[0]);
        }
        @Override
        public String toString() {
            return "StatusEvent" + "[type=" + type
                 + ",humanCode=" + humanCode + ",channelType=" + channelType
                 + ",status=" + status +"]";
        }

        public Contact.Status status;
    }

    public class RequestEvent extends EventArgs {
        public RequestEvent(int type, String humanCode, int channelType, byte[] data) {
            super(type, humanCode, channelType, data);
            summary = new String(data);
        }
        @Override
        public String toString() {
            return "RequestEvent" + "[type=" + type
                 + ",humanCode=" + humanCode + ",channelType=" + channelType
                 + ",summary=" + summary +"]";
        }

        public String summary;
    }

    public class InfoEvent extends EventArgs {
        public InfoEvent(int type, String humanCode, int channelType, byte[] data) {
            super(type, humanCode, channelType, data);

            String info = new String(data);
            if(info.contains(JsonKey.IsMyself) == true) {
                UserInfo userInfo = new ContactInterface.UserInfo();
                userInfo.fromJson(info);
                humanInfo = userInfo;
            } else if(info.contains(JsonKey.IsFriend) == true) {
                FriendInfo friendInfo = new ContactInterface.FriendInfo();
                friendInfo.fromJson(info);
                humanInfo = friendInfo;
            } else {
               Log.w(Utils.TAG, "InfoEvent: Failed to parse human data.");
            }
        }
        @Override
        public String toString() {
            return "InfoEvent" + "[type=" + type
                 + ",humanCode=" + humanCode + ",channelType=" + channelType
                 + ",humanInfo=" + humanInfo +"]";
        }

        public ContactInterface.HumanInfo humanInfo;
    }

    public class MsgAckEvent extends EventArgs {
        public MsgAckEvent(int type, String humanCode, int channelType, byte[] data) {
            super(type, humanCode, channelType, data);

            JsonObject msgAckEvent = Utils.GetGsonBuilder().create().fromJson(new String(data), JsonObject.class);
            this.memo = msgAckEvent.get(JsonKey.Memo).getAsString();
            this.ackToNanoTime = msgAckEvent.get(JsonKey.AckToNanoTime).getAsLong();
            //this.nanoTime = Long.reverseBytes(new BigInteger(data).longValue());
        }
        @Override
        public String toString() {
            return "MsgAckEvent" + "[type=" + type
                 + ",humanCode=" + humanCode + ",channelType=" + channelType
                 + ",memo=" + memo
                 + ",ackToNanoTime=" + ackToNanoTime +"]";
        }

        public String memo;
        public long ackToNanoTime;
    }

    protected ContactListener() {
        super(ContactListener.class.getName(), 0);
    }

    @CrossInterface
    private byte[] onAcquire(int reqType, String pubKey, byte[] data, String extra) {
        Log.i(Utils.TAG, "ContactListener.onAcquire()");

        AcquireArgs args = new AcquireArgs(reqType, pubKey, data, extra);
        byte[] ret = onAcquire(args);

        return ret;
    }

    @CrossInterface
    private void onEvent(int eventType, String humanCode, int channelType, byte[] data) {

        EventArgs args = null;

        EventArgs.Type type = EventArgs.Type.valueOf(eventType);
        switch (type) {
            case StatusChanged:
                args = new StatusEvent(eventType, humanCode, channelType, data);
                break;
            case FriendRequest:
                args = new RequestEvent(eventType, humanCode, channelType, data);
                break;
            case HumanInfoChanged:
                args = new InfoEvent(eventType, humanCode, channelType, data);
                break;
            case MessageAck:
                args = new MsgAckEvent(eventType, humanCode, channelType, data);
                break;
            default:
                throw new RuntimeException("Unimplemented type: " + type);
        }

        Log.i(Utils.TAG, "ContactListener.onEvent() args=" + args);
        onEvent(args);
        return;
    }

    @CrossInterface
    private void onReceivedMessage(String humanCode, int channelType,
                                   int type, byte[] data, String cryptoAlgorithm, String memo,
                                   long nanoTime, long replyToNanoTime) {
        Contact.Message message = new Contact.Message(ContactMessage.Type.valueOf(type),
                                                      data, cryptoAlgorithm, memo);
        message.nanoTime = nanoTime;
        message.replyToNanoTime = replyToNanoTime;

        onReceivedMessage(humanCode, ContactChannel.valueOf(channelType), message);
        return;
    }
}
